File: java-backend/src/test/java/com/study/service/ArticleServiceTest.java
package com.study.service;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.study.cache.RedisArticleCache;
import com.study.domain.Article;
import com.study.domain.User;
import com.study.exception.ForbiddenException;
import com.study.exception.ResourceNotFoundException;
import com.study.mapper.ArticleMapper;
import com.study.mapper.UserMapper;
import com.study.storage.StorageService;
import com.study.web.dto.ArticleRequest;
import com.study.web.dto.ArticleResponse;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.context.ApplicationEventPublisher;

/** ArticleService 单元测试:创建、作者归属校验、未找到、删除。 */
@ExtendWith(MockitoExtension.class)
class ArticleServiceTest {

    @Mock
    ArticleMapper articleMapper;
    @Mock
    UserMapper userMapper;
    @Mock
    StorageService storageService;
    @Mock
    RedisArticleCache cache;
    @Mock
    ApplicationEventPublisher eventPublisher;
    @InjectMocks
    ArticleService articleService;

    @Test
    void create_returnsResponseWithAuthorAndZeroViews() {
        when(articleMapper.insert(any(Article.class))).thenReturn(1);
        User author = new User();
        author.setId(1L);
        author.setUsername("alice");
        when(userMapper.selectById(1L)).thenReturn(author);

        ArticleResponse response = articleService.create(new ArticleRequest("t", "c", "cat", null), 1L);

        assertThat(response.title()).isEqualTo("t");
        assertThat(response.authorId()).isEqualTo(1L);
        assertThat(response.authorUsername()).isEqualTo("alice");
        assertThat(response.viewCount()).isEqualTo(0L);
    }

    @Test
    void update_byNonOwner_throwsForbidden() {
        Article article = new Article();
        article.setId(10L);
        article.setAuthorId(1L);
        when(articleMapper.selectById(10L)).thenReturn(article);

        assertThatThrownBy(() -> articleService.update(10L, new ArticleRequest("t2", "c2", null, null), 2L))
                .isInstanceOf(ForbiddenException.class);
    }

    @Test
    void get_missing_throwsNotFound() {
        when(cache.get(99L)).thenReturn(null);
        when(articleMapper.selectById(99L)).thenReturn(null);

        assertThatThrownBy(() -> articleService.get(99L))
                .isInstanceOf(ResourceNotFoundException.class);
    }

    @Test
    void delete_byOwner_deletesAndEvicts() {
        Article article = new Article();
        article.setId(5L);
        article.setAuthorId(1L);
        when(articleMapper.selectById(5L)).thenReturn(article);
        when(articleMapper.deleteById(5L)).thenReturn(1);

        articleService.delete(5L, 1L);

        verify(articleMapper).deleteById(5L);
        verify(cache).evict(5L);
    }
}

File: kotlin-backend/src/test/kotlin/com/study/service/ArticleServiceTest.kt
package com.study.service

import com.study.cache.RedisArticleCache
import com.study.domain.Article
import com.study.domain.User
import com.study.exception.ForbiddenException
import com.study.exception.ResourceNotFoundException
import com.study.mapper.ArticleMapper
import com.study.mapper.UserMapper
import com.study.storage.StorageService
import com.study.web.dto.ArticleRequest
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import org.springframework.context.ApplicationEventPublisher

/** ArticleService 单元测试(MockK):创建、作者归属校验、未找到、删除。 */
class ArticleServiceTest {

    private val articleMapper = mockk<ArticleMapper>()
    private val userMapper = mockk<UserMapper>()
    private val storageService = mockk<StorageService>()
    private val cache = mockk<RedisArticleCache>(relaxed = true)
    private val eventPublisher = mockk<ApplicationEventPublisher>(relaxed = true)
    private val articleService = ArticleService(articleMapper, userMapper, storageService, cache, eventPublisher)

    @Test
    fun `创建 返回含作者与0浏览量`() {
        every { articleMapper.insert(any<Article>()) } returns 1
        every { userMapper.selectById(1L) } returns User().apply {
            id = 1L
            username = "alice"
        }

        val resp = articleService.create(ArticleRequest("t", "c", "cat", null), 1L)

        assertEquals("t", resp.title)
        assertEquals(1L, resp.authorId)
        assertEquals("alice", resp.authorUsername)
        assertEquals(0L, resp.viewCount)
    }

    @Test
    fun `非作者更新 抛禁止`() {
        val article = Article().apply {
            id = 10L
            authorId = 1L
        }
        every { articleMapper.selectById(10L) } returns article

        assertThrows(ForbiddenException::class.java) {
            articleService.update(10L, ArticleRequest("t2", "c2", null, null), 2L)
        }
    }

    @Test
    fun `详情不存在 抛未找到`() {
        every { cache.get(99L) } returns null
        every { articleMapper.selectById(99L) } returns null

        assertThrows(ResourceNotFoundException::class.java) { articleService.get(99L) }
    }

    @Test
    fun `作者删除 执行删除并失效缓存`() {
        val article = Article().apply {
            id = 5L
            authorId = 1L
        }
        every { articleMapper.selectById(5L) } returns article
        every { articleMapper.deleteById(5L) } returns 1

        articleService.delete(5L, 1L)

        verify { articleMapper.deleteById(5L) }
        verify { cache.evict(5L) }
    }
}
